woocommerce/woocommerce
# Store API: Add support to query product_variations by attributes

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



Currently you cannot query product variations by attributes using the Store API, despite the [documentation saying that it is possible](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/products.md#product-variations).

In WooCommerce, when querying product variations specifically, the `tax_query` is not used directly because product attributes (like size, color, etc.) associated with variations are stored as post meta, not as taxonomy terms. Hence, you need to use `meta_query` instead of `tax_query` when dealing with product variations. Product variations are considered as individual products tied to the main product. These variations hold their specific attributes as meta values in the database, rather than being treated as taxonomy terms.

This PR introduces a method which converts the `tax_query` into a `meta_query` for types `product_variation` only.

**Known limitation**: This will still return no results if you're filtering by attribute ID, in order for this to work you need to pass the attribute slug.



Closes https://github.com/woocommerce/woocommerce/issues/42093



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Using the sample products**: [sample_products.csv](https://github.com/woocommerce/woocommerce/files/13725228/sample_products.csv)

1. Confirm issue prior without this PR beforehand with the following request (you can use Postman or any other client): `/wp-json/wc/store/v1/products?attributes[0][attribute]=pa_color&attributes[0][slug]=red&type=variation` This should return no results.
2. Apply this PR and rerun the request.
3. Confirm results returned as as expected.
4. Try querying products instead of product variations (by removing `type=variation`) to confirm this still works as expected and no regression has been introduced.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Store API: Add support to query `product_variations` by attribute slugs.

#### Comment 

</details>
